sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor.cs

```diff
@@ -590,24 +592,64 @@ LastEnqueuedEventProperties readLastEnquedEventInformation()
                 return new LastEnqueuedEventProperties(consumer.LastReceivedEvent);
             }
 
+            Logger.EventProcessorPartitionProcessingStart(partitionId, Identifier, EventHubName, ConsumerGroup);
+
             // Define the routine to handle processing for the partition.
 
             async Task performProcessing()
             {
-                cancellationSource.Token.ThrowIfCancellationRequested<TaskCanceledException>();
-
+                cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
                 var connection = default(EventHubConnection);
                 var retryDelay = default(TimeSpan?);
                 var capturedException = default(Exception);
                 var eventBatch = default(IReadOnlyList<EventData>);
                 var lastEvent = default(EventData);
                 var failedAttemptCount = 0;
                 var failedConsumerCount = 0;
+                var startingPosition = Options.DefaultStartingPosition;
 
                 // Create the connection to be used for spawning consumers; if the creation
                 // fails, then consider the processing task to be failed.  The main processing
                 // loop will take responsibility for attempting to restart or relinquishing ownership.
 
+                try
+                {
+                    // Initialize the partition context; the handler is responsible for initialing any custom fields of the partition type.
+
+                    await OnInitializingPartitionAsync(partition, cancellationToken).ConfigureAwait(false);
```

Comment:
It may be helpful to abstract out the initialization/checkpoint bits into a virtual method that we can delegate to.   That would reduce the complexity here, make clear the split in responsibilities (initialization / processing),  allow you to mock initialization out for testing the processing, and test initialization in isolation.